src/GraphQL/Federation/Resolvers/EntityResolver.cs

```diff
@@ -0,0 +1,213 @@
+                //  deserialization mapping
+                //value = rep.ToObject(resolver.SourceType, null!);
+                if (resolver.SourceType == typeof(object) || resolver.SourceType == typeof(Dictionary<string, object>) || resolver.SourceType == typeof(IDictionary<string, object?>))
+                    value = rep;
+                else
+                    value = ToObject(resolver.SourceType, objectGraphType, rep);
+            }
+            catch (Exception ex)
+            {
+                // mask the underlying exception to prevent leaking implementation details
+                // the InnerException can be read for debugging purposes
+                throw new InvalidOperationException($"Error converting representation for type '{typeName}'.", ex);
+            }
+
+            ret.Add(new Representation(objectGraphType, resolver, value));
+        }
+        return ret;
+    }
+
+    /// <summary>
+    /// Deserializes an object based on properties provided in a dictionary, using graph type information from
+    /// an output graph type. Requires that the object type has a parameterless constructor.
+    /// </summary>
+    private static object ToObject(Type objectType, IObjectGraphType objectGraphType, IDictionary<string, object?> map)
+    {
+        // create an instance of the target CLR type
+        var obj = Activator.CreateInstance(objectType)!;
+
+        // loop through each field in the map and deserialize the value to the corresponding property on the object
+        foreach (var item in map)
+        {
+            // skip the __typename field as it was already used to find the object graph type and is not intended to be deserialized
+            if (item.Key == "__typename")
+                continue;
+
+            // find the field on the object graph type, and the corresponding property on the object type
+            var field = objectGraphType.Fields.Find(item.Key)
+                ?? throw new InvalidOperationException($"Field '{item.Key}' not found in graph type '{objectGraphType.Name}'.");
+            var graphType = field.ResolvedType!;
+            var prop = objectType.GetProperty(item.Key, System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.IgnoreCase | System.Reflection.BindingFlags.Public)
+                ?? throw new InvalidOperationException($"Property '{item.Key}' not found in type '{objectType.GetFriendlyName()}'.");
+
+            // deserialize the value
+            var value = Deserialize(item.Key, graphType, prop.PropertyType, item.Value);
+            // convert the value to the property type if necessary using the ValueConverter (typical for ID fields that convert to numbers)
+            if (value != null && !prop.PropertyType.IsInstanceOfType(value))
+                value = ValueConverter.ConvertTo(value, prop.PropertyType);
+            // set the property value
+            prop.SetValue(obj, value);
+        }
+        return obj;
```

Comment:
This reflection may be expensive... Is it possible to rewrite with compiled expressions similar to input types?

Reply:
I want to for sure .... but I think it's more important to get the codebase functional, and then we can work on performance.  Ideally I'd like to re-use the input object graph type functionality so we are not duplicating a whole bunch of code.  But that's easier said than done.

Reply:
Makes sense